### PR TITLE
Added FlexibleAutoExt mode for CKKS.

### PR DESCRIPTION
**Pull request by sarojaduality**

### CODE REVIEW

File: src/pke/lib/encoding/ckkspackedencoding.cpp [hunk 1]
```diff
@@ -374,6 +374,160 @@ bool CKKSPackedEncoding::Encode() {
...
+    usint numTowers = nativeParams.size();
+    std::vector<DCRTPoly::Integer> moduli(numTowers);
+    for (usint i = 0; i < numTowers; i++) {
+      moduli[i] = nativeParams[i]->GetModulus();
+    }
+
+    int32_t logc = 0, logci;
+    for (size_t i = 0; i < Nh; ++i) {
+      inverse[i] *= scalingFactor;
+      logci = static_cast<int32_t>(ceil(log2(abs(inverse[i].real()))));
```
Comment (dsuponitskiy) [inline src/pke/lib/encoding/ckkspackedencoding.cpp:402]: logci should be defined here instead of line 399

**Review by dsuponitskiy — COMMENTED** [review]
Saroja, please let me know if you want to discuss my comments and when

File: src/pke/lib/scheme/ckksrns/ckksrns-parametergeneration.cpp [hunk 2]
```diff
@@ -100,6 +121,18 @@ bool ParameterGenerationCKKSRNS::ParamsGenCKKSRNS(
 
       // Choose ring dimension based on security standards
       n = nRLWE(qBound);
+      if (extraModSize > 0) {
+        usint nExact = nRLWE(qBoundExact);
+        //std::cerr << "n = " << n << std::endl;
+        //std::cerr << "nExact = " << nExact << std::endl;
+        while (n > nExact) {
```
[the remark this answers is not in the document]
Comment (sarojaduality) [reply src/pke/lib/scheme/ckksrns/ckksrns-parametergeneration.cpp:128]: You're right, good catch. @yspolyakov @kimandrik I ported this code from the _andre-ckks_ branch of palisade-development. This part of the code is supposed to ensure the security level is met. Could you point me to a reference for these security standards, so I can double check this piece of code?

File: src/pke/lib/scheme/ckksrns/ckksrns-parametergeneration.cpp [hunk 3]
```diff
@@ -123,9 +156,15 @@ bool ParameterGenerationCKKSRNS::ParamsGenCKKSRNS(
 
   usint dcrtBits = scaleExp;
 
-  std::vector<NativeInteger> moduliQ(numPrimes);
-  std::vector<NativeInteger> rootsQ(numPrimes);
-
+  std::vector<NativeInteger> moduliQ;
```
[the remark this answers is not in the document]
Comment (sarojaduality) [reply src/pke/lib/scheme/ckksrns/ckksrns-parametergeneration.cpp]: Fixed

File: src/pke/lib/scheme/ckksrns/ckksrns-parametergeneration.cpp [hunk 4]
```diff
@@ -78,14 +84,29 @@ bool ParameterGenerationCKKSRNS::ParamsGenCKKSRNS(
...
 
+  uint32_t qBoundExact = 0;
+  uint32_t auxBitsExact = 0;
+
+  if (rsTech == FLEXIBLEAUTOEXT) {
+      qBoundExact = firstModSize + (numPrimes - 1) * scaleExp;
+    if (ksTech == BV) {
+      qBoundExact += ceil(static_cast<double>(qBoundExact) / auxBits) * auxBits;
+    } else if (ksTech == HYBRID) {
+      auxBitsExact = ceil(ceil(static_cast<double>(qBoundExact) / numPartQ) /
```
[the remark this answers is not in the document]
Comment (sarojaduality) [reply src/pke/lib/scheme/ckksrns/ckksrns-parametergeneration.cpp]: Fixed

File: src/pke/lib/scheme/ckksrns/ckksrns-parametergeneration.cpp [hunk 5]
```diff
@@ -78,14 +84,29 @@ bool ParameterGenerationCKKSRNS::ParamsGenCKKSRNS(
   uint32_t qBound = 0;
```
[the remark this answers is not in the document]
Comment (sarojaduality) [reply src/pke/lib/scheme/ckksrns/ckksrns-parametergeneration.cpp]: Fixed

File: src/pke/lib/encoding/ckkspackedencoding.cpp [hunk 6]
```diff
@@ -374,6 +374,160 @@ bool CKKSPackedEncoding::Encode() {
...
+    BigVector temp(this->GetElementRingDimension(), this->GetElementModulus());
+
+    int64_t re, im;
+    size_t i, jdx, idx;
+    double dre, dim;
+    for (i = 0, jdx = Nh, idx = 0; i < Nh; ++i, jdx++, idx++) {
+      dre = inverse[i].real() * powP;
+      dim = inverse[i].imag() * powP;
+      // Check for possible overflow in llround function
+      if (std::abs(dre) >= dq || std::abs(dim) >= dq) {
```
[the remark this answers is not in the document]
Comment (sarojaduality) [reply src/pke/lib/encoding/ckkspackedencoding.cpp]: Fixed to compare doubles with epsilon and fabs

File: src/pke/lib/encoding/ckkspackedencoding.cpp [hunk 7]
```diff
@@ -374,6 +374,160 @@ bool CKKSPackedEncoding::Encode() {
...
+    double powP = pow(2, p * depth);
+
+    const BigInteger &q = this->GetElementModulus();
+    // min of q and 2^63-2^9-1 - max value
+    // that could be round to int64_t
+    double dq = std::min(9223372036854775295., q.ConvertToDouble());
+
+    BigVector temp(this->GetElementRingDimension(), this->GetElementModulus());
+
+    int64_t re, im;
```
[the remark this answers is not in the document]
Comment (sarojaduality) [reply src/pke/lib/encoding/ckkspackedencoding.cpp]: fixed

File: src/pke/lib/encoding/ckkspackedencoding.cpp [hunk 8]
```diff
@@ -374,6 +374,160 @@ bool CKKSPackedEncoding::Encode() {
...
+  } else if (this->typeFlag == IsNativePoly) {
+    double p = this->encodingParams->GetPlaintextModulus();
+    double powP = pow(2, p * depth);
+
+    int64_t q;
+    q = this->GetElementModulus().ConvertToInt();
+    NativeVector temp(this->GetElementRingDimension(), q);
+
+    double dq = q;
+    size_t i, jdx, idx;
```
[the remark this answers is not in the document]
Comment (sarojaduality) [reply src/pke/lib/encoding/ckkspackedencoding.cpp]: fixed

File: src/pke/lib/encoding/ckkspackedencoding.cpp [hunk 9]
```diff
@@ -374,6 +374,160 @@ bool CKKSPackedEncoding::Encode() {
...
+
+    if (logc < 0) {
+      OPENFHE_THROW(math_error, "Too small scaling factor");
+    }
+
+    int32_t logValid = (logc <= 62) ? logc : 62;
+    int32_t logApprox = logc - logValid;
+    double approxFactor = pow(2, logApprox);
+    std::vector<int64_t> temp(N);
+    size_t i, jdx, idx;
```
[the remark this answers is not in the document]
Comment (sarojaduality) [reply src/pke/lib/encoding/ckkspackedencoding.cpp]: fixed

File: src/pke/lib/encoding/ckkspackedencoding.cpp [hunk 10]
```diff
@@ -374,6 +374,160 @@ bool CKKSPackedEncoding::Encode() {
...
+      dim = inverse[i].imag() / approxFactor;
+
+      re = static_cast<int64_t>(dre + 0.5);
+      im = static_cast<int64_t>(dim + 0.5);
+
+      temp[idx] = (re < 0) ? q + re : re;
+      temp[jdx] = (im < 0) ? q + im : im;
+    }
+
+    for (i = 0; i < nativeParams.size(); i++) {
```
[the remark this answers is not in the document]
Comment (sarojaduality) [reply src/pke/lib/encoding/ckkspackedencoding.cpp]: fixed

File: src/pke/lib/encoding/ckkspackedencoding.cpp (hunk 1 above)
Comment (sarojaduality) [reply src/pke/lib/encoding/ckkspackedencoding.cpp]: fixed

File: src/pke/lib/encoding/ckkspackedencoding.cpp [hunk 11]
```diff
@@ -374,6 +374,160 @@ bool CKKSPackedEncoding::Encode() {
 }
 #endif
 
+bool CKKSPackedEncoding::EncodeWithExtra() {
```
Comment (yspolyakov) [inline src/pke/lib/encoding/ckkspackedencoding.cpp:377]: @sarojaduality The logic for `EncodeWithExtra` seems to unnecessarily complicated.  For `Encode`, we only support DCRTPoly. We should support only DCRTPoly here, too. 

1. Please remove the two other branches from `EncodeWithExtra`. 
2. Please review the implementation of `Encode` (the logic should the same for `Encode` and `EncodeWithExtra`, except for the extra part). Maybe `EncodeWithExtra` can be combined with `Encode`.

File: src/pke/lib/scheme/ckksrns/ckksrns-parametergeneration.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (yspolyakov) [reply src/pke/lib/scheme/ckksrns/ckksrns-parametergeneration.cpp:124]: @kimandrik I don't understand what this loop. I understand the intent, but PBound seems small (like 20 bits). What are we actually doing here?

File: src/pke/lib/encoding/ckkspackedencoding.cpp (hunk 11 above)
Comment (sarojaduality) [reply src/pke/lib/encoding/ckkspackedencoding.cpp:414]: Removed the EncodeWithExtra function, so we only use Encode instead.

File: src/pke/lib/scheme/ckksrns/ckksrns-parametergeneration.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (yspolyakov) [reply src/pke/lib/scheme/ckksrns/ckksrns-parametergeneration.cpp:124]: @kimandrik Could you review this code? I don't think it does anything useful

**Pull request merged**